### PR TITLE
Implement wasi-config

### LINKED ISSUE #2868: Support wasi-config
WASI is moving wasi-config to phase 3 soon. We should consider moving from the [spin variables](https://github.com/fermyon/spin/blob/main/wit/variables.wit) interface to [WASI config](https://github.com/WebAssembly/wasi-runtime-config).

The main diff is the need to add support for `get-all`

Spin 3.0 may be a good opportunity to make this switch

### PR DESCRIPTION
**Pull request by itowlson**
Fixes #2868.

This has the same considerations as our wasi-observe and wasi-keyvalue implementations in terms of versioning the Spin worlds - I mashed it into the existing Spin 2 worlds for development and testing but know that is not sustainable.  (At this point, I'm assuming all this lot will be punted into a new Spin 3 world.)

I'm publishing this mostly to:

1. see if it works
2. give us the chance to look at implementation considerations and decide if we need to offer feedback to upstream proposal

(Regarding 2, it felt like we have some errors that don't map nicely to the proposal.  I ended up using the `io` error variant for anything that wasn't a provider error, which feels potentially misleading.  @kate-goldenring I'd value your thoughts on this!)

### CODE REVIEW

File: crates/expressions/src/lib.rs
Comment (rylev) [inline crates/expressions/src/lib.rs:55]: Nit: doc comment.

File: crates/expressions/src/lib.rs
Comment (rylev) [inline crates/expressions/src/lib.rs:67]: I am slightly worried about performance - at some point we might want to see if adding some sort of caching mechanism to providers is worth it.

File: crates/factor-variables/src/host.rs
Comment (rylev) [inline crates/factor-variables/src/host.rs:49]: FWIW you could:

```rust
Err(variables::Error::Undefined(_)) => Ok(None)
```

File: crates/factor-variables/src/lib.rs
Comment (lann) [reply crates/factor-variables/src/lib.rs:34]: Crate feature flag?

File: crates/expressions/src/lib.rs
Comment (lann) [reply crates/expressions/src/lib.rs:67]: We should probably push this down into the provider as e.g. (untested):

```rust
trait Provider {
    async fn get(&self, key: &Key) -> anyhow::Result<Option<String>>;

    async fn get_many(&self, keys: impl IntoIterator<Item = &Key>) -> anyhow::Result<impl IntoIterator<Option<String>>> {
        try_join_all(keys.into_iter().map(|key| self.get(key)).await
    }
}
```

File: tests/test-components/components/wasi-config/src/lib.rs
Comment (kate-goldenring) [reply tests/test-components/components/wasi-config/src/lib.rs:17]: No it doesn't: https://github.com/WebAssembly/wasi-runtime-config/blob/f4d699bc6dd77adad99fa1a2246d482225ec6485/wit/store.wit#L17

This is a good call out of how it should probably be opinionated about string fmt or support `invalid-name` error

File: crates/factor-variables/src/host.rs
Comment (itowlson) [reply crates/factor-variables/src/host.rs:82]: @kate-goldenring Yes, that's exactly my worry.  I appreciate that ""other" is a kind of unhelpful bucket though...!

**Comment by kate-goldenring** [comment]
> @kate-goldenring re [#2869 (comment)](https://github.com/fermyon/spin/pull/2869#discussion_r1781470115) can you clarify "no it doesn't" please? I didn't follow how the linked WIT comment applied - are you thinking of this as "we can't possibly find a key for an invalid name so should return `Ok(None)`" kind of thing?

@itowlson I was aiming to point out that this is a weak spot of the WASI config interface and that we should propose updating it to either (A) be opinionated about string formatting or (B) support and invalid key error. My preference is that latter; however, the former would help achieve @rylev's point of enabling more portability across host implementatations.

File: crates/factor-variables/src/lib.rs
Comment (rylev) [reply crates/factor-variables/src/lib.rs:34]: Seems like the feature is light-weight enough that we could just do a dynamic check instead of a compile time feature flag, but I don't feel strongly either way.

**Comment by itowlson** [comment]
Updated this to sit over the world3 proposal (#2887).  The wasi-config stuff is only the _final_ commit - the first three are the world3 proposal baseline.

File: crates/factor-variables/src/lib.rs
Comment (itowlson) [reply crates/factor-variables/src/lib.rs:34]: We'll need to make a decision on this (and a similar consideration for wasi-keyvalue) fairly soon if we hope to land this in Spin 3.

Comment (itowlson) [reply crates/factor-variables/src/lib.rs:34]: It sounds like you're expressing a general principle about allowing embedders to control which interfaces get linked, rather than anything specific to wasi-config, or am I misunderstanding?

File: tests/test-components/components/wasi-config/README.md
Comment (itowlson) [reply tests/test-components/components/wasi-config/README.md:1]: I can't do that as part of this PR but I've raised https://github.com/fermyon/conformance-tests/issues/40 to remind us to do so once this merges.

File: crates/factor-variables/src/host.rs
Comment (itowlson) [reply crates/factor-variables/src/host.rs]: Conclusion was for invalid names to be treated as not found

File: crates/expressions/src/lib.rs
Comment (itowlson) [reply crates/expressions/src/lib.rs:68]: I don't think any of our current providers can optimise this beyond running the futures concurrently - all appear to support retrieving only one secret at once - so I propose we defer this.

Comment (kate-goldenring) [reply crates/expressions/src/lib.rs:68]: @itowlson instead of adding this function, could we use lann's suggestion from [`get_all`](https://github.com/fermyon/spin/pull/2869/files#r1778592084)

**Review by kate-goldenring — APPROVED** [review]
Thank you for adding this and doing all the work to assemble a world for Spin 3.0!

**Comment by itowlson** [comment]
@kate-goldenring I'm sceptical about pushing `get_all` down to the provider level because:

1. We would still need to do a `try_join_all` at the top level, because there could be multiple providers in play, even if each provider was batching. 
2. If we sent a `get_all` off to each provider, the code for "rejoining" them would become more complicated, because two providers might return values for the same key, so we would need to take account of precedence (which `resolve_template` already does for us).
3. There appears to be no performance benefit for any current provider because none of the underlying implementations can batch requests.

Or maybe I'm misunderstanding the suggestion.  Wouldn't be the first time...

If we're looking to optimise then I think the most practical thing we can do is retain a cache for the duration of the get-all request so we can go "oh I need to resolve key `blah`, but I've resolved that already in this cycle, so I'll just use the same value."  Given timescales, my preference would be that we don't block on that and instead consider it as a non-breaking change in a subsequent PR, either before or after 3.0 depending on how we go.  But I can take a look if folks prefer.

**Pull request merged**
